FreeRTOS_ARP.c

```diff
@@ -119,27 +141,103 @@ eFrameProcessingResult_t eARPProcessPacket( ARPPacket_t * const pxARPFrame )
 
     pxARPHeader = &( pxARPFrame->xARPHeader );
 
-    /* Only Ethernet hardware type is supported.
-     * Only IPv4 address can be present in the ARP packet.
-     * The hardware length (the MAC address) must be 6 bytes. And,
-     * The Protocol address length must be 4 bytes as it is IPv4. */
-    if( ( pxARPHeader->usHardwareType == ipARP_HARDWARE_TYPE_ETHERNET ) &&
-        ( pxARPHeader->usProtocolType == ipARP_PROTOCOL_TYPE ) &&
-        ( pxARPHeader->ucHardwareAddressLength == ipMAC_ADDRESS_LENGTH_BYTES ) &&
-        ( pxARPHeader->ucProtocolAddressLength == ipIP_ADDRESS_LENGTH_BYTES ) )
+    /* The field ucSenderProtocolAddress is badly aligned, copy byte-by-byte. */
+
+    /*
+     * Use helper variables for memcpy() to remain
+     * compliant with MISRA Rule 21.15.  These should be
+     * optimized away.
+     */
+    pvCopySource = pxARPHeader->ucSenderProtocolAddress;
+    pvCopyDest = &ulSenderProtocolAddress;
+    ( void ) memcpy( pvCopyDest, pvCopySource, sizeof( ulSenderProtocolAddress ) );
+    /* The field ulTargetProtocolAddress is well-aligned, a 32-bits copy. */
+    ulTargetProtocolAddress = pxARPHeader->ulTargetProtocolAddress;
+
+    if( uxARPClashCounter != 0 )
     {
-        /* The field ulSenderProtocolAddress is badly aligned, copy byte-by-byte. */
-
-        /*
-         * Use helper variables for memcpy() to remain
-         * compliant with MISRA Rule 21.15.  These should be
-         * optimized away.
-         */
-        pvCopySource = pxARPHeader->ucSenderProtocolAddress;
-        pvCopyDest = &ulSenderProtocolAddress;
-        ( void ) memcpy( pvCopyDest, pvCopySource, sizeof( ulSenderProtocolAddress ) );
-        /* The field ulTargetProtocolAddress is well-aligned, a 32-bits copy. */
-        ulTargetProtocolAddress = pxARPHeader->ulTargetProtocolAddress;
+        /* Next defensive request must not be sent for arpIP_CLASH_RESET_TIMEOUT_MS
+         * period. */
+        TickType_t xARPClashTimeout = pdMS_TO_TICKS( arpIP_CLASH_RESET_TIMEOUT_MS );
+
+        /* Has the timeout been reached? */
+        if( xTaskCheckForTimeOut( &ARPClashTimeOut, &xARPClashTimeout ) == pdTRUE )
```

Comment:
Could xTaskCheckForTimeOut() be called before vTaskSetTimeOutState()?

Reply:
I don't think so. However that condition will occur since `uxARPClashCounter` will never be non-zero unless `vTaskSetTimeOutState` has been already called.
Also, since this is in the same function - there is no race condition.